### PR TITLE
[fx][split_module]Optimize inference-purposing graph splitting with lazy GraphModule construction

### PR DESCRIPTION
**Pull request by fxdawnn**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #179839

## Background

This work started from a practical problem in vllm's `torch.compile` pipeline: `torch.Size` objects cannot cross FX graph split boundaries because `aot_autograd` can't handle them as submodule inputs/outputs. vllm had a custom `_vllm_split_module` that worked around this, but the fix belonged upstream in PyTorch.

The initial contribution was `_decompose_size_nodes()` — a pre-pass that replaces `x.size()` calls with individual `sym_size.int(x, dim)` nodes before splitting, so only scalar SymInts (not Size tuples) cross partition boundaries. This was paired with `split_module_simple`, originally a lean 2-pass alternative to `split_module` that skips autocast/grad handling and topological sort for inference-only Dynamo graphs.

Early benchmarks of the 2-pass algorithm showed only **~7-10% improvement** over `split_module` — far less than expected. Investigation revealed the actual bottleneck was not graph traversal or node cloning, but **`GraphModule` construction overhead** when creating 70-100+ partition submodules. Each `GraphModule` triggers:
- A per-instance Python class creation (`GraphModule.__new__` with MRO traversal)
- 17+ container allocations via `nn.Module.__init__`
- Node iteration to find `get_attr`/`call_module` targets
- `recompile()` codegen generating Python source via `compile()` + `exec()`

For parameterless partition submodules that only serve as graph containers, all of this is wasted. The final optimization bypasses this entirely by directly populating `__dict__` and using `_LazyGraphModule._lazy_forward` for deferred codegen, achieving **5.5x speedup** on the split_graph phase.

## Summary

Adds `split_module_simple()` and `_decompose_size_nodes()` to `torch/fx/passes/split_module.py`.

## Changes

**`torch/fx/passes/split_module.py`:**
- `split_module_simple(m, node_to_partition, *, partition_affix=None)` — 2-pass graph splitter that takes a pre-computed partition dict instead of a callback. Skips autocast/grad handling, get_attr special-casing, and topological sort (not needed for inference graphs).
- `_decompose_size_nodes(m)` — pre-pass that decomposes `x.size()` into per-dim `sym_size.int(x, dim)` calls. `torch.Size` objects cannot cross split boundaries because `aot_autograd` cannot handle them as submodule outputs.
- Lightweight `_make_lite_gm` inner function that constructs `GraphModule` instances by bypassing `__new__`, `nn.Module.__init__`, `GraphModule.__init__`, and `recompile()`.

**`test/test_fx_experimental.py`:**
- 6 new tests: basic split, multi-output, single partition, partition affix, symint crossing, equivalence with `split_module`.

## Benchmark Results

Measured on vllm model serving with `TORCH_TRACE` profiling (cold cache):

| Model | `split_module` | `split_module_simple` | Speedup |
|-------|---------------|----------------------|---------|
| GLM-4.7-FP8 (8x GPU, 105 partitions) | 848 ms | 159 ms | **5.3x** |
| Llama (4x GPU, 81 partitions) | 479 ms | 87 ms | **5.5x** |
| GPT(openai/gpt-oss-120b) (4x GPU, 73 partitions) | 210 ms | 38 ms | **5.5x** |

No downstream regression — AOTAutograd cache deduplication is preserved (identical subgraph counts and cache hit rates across baseline and `split_module_simple`).

## Investigation Journey: Finding the Real Bottleneck

We initially assumed the bottleneck in `split_module` was the per-node cloning work — `map_arg` callbacks, `Graph.create_node`, `node.meta.copy()` — since `split_module_simple` reduces graph passes from 5 to 2. However, early benchmarks showed only **~7-10% improvement**, far less than the expected 2x.

**Phase 1: Python-level `split_module_simple` (v1)**
A clean 2-pass Python implementation gave ~7-10% improvement. Profiling the node clone loop showed each 86-node partition clone took only ~0.2ms. With 73 partitions, total cloning was ~12ms out of ~190ms. The clone loop was never the bottleneck.

**Phase 2: C++ fused node cloning (v3-v4)**
We added `_fx_node_copy` and `_fx_bulk_clone_nodes` to `torch/csrc/fx/node.cpp` — fusing `map_arg` remapping + `_update_args_kwargs` edge-building into a single `map_aggregate` pass with C++ `PyDict_GetItemWithError` lookups instead of Python callbacks. This eliminated per-node Python-C++ boundary crossings. Result: **<1% additional improvement**. The per-node C++ overhead was already fast; the Python loop overhead was negligible.

**Phase 3: Phase-level profiling reveals the truth**
Adding timing to each phase of `split_module_simple` showed:
```
dep_detect=10ms  placeholders=5ms  clone=9ms  stitch=82ms  make_gm=82ms  total=190ms
```
**`stitch` (82ms) + `make_gm` (82ms) = 87% of total time.** These were the `_make_graph_module()` calls — 73 for partition submodules + 1 for the outer stitching module. The C++ optimizations were targeting ~9ms (5%) of the total.

**Phase 4: Lightweight GraphModule construction**
We bypassed `GraphModule.__new__`, `nn.Module.__init__`, and `GraphModule.__init__` for partition submodules, directly populating `__dict__` entries. Initial attempt still showed no improvement because `recompile()` was being called — the `_use_lazy_graph_module` flag wasn't active in the vllm backend context. Using `_LazyGraphModule` directly (instead of relying on the global flag) eliminated `recompile()` codegen and dropped the total from 190ms to **30ms**.

**Phase 5: AOTAutograd cache dedup fix**
Benchmarking revealed that `split_module_simple` broke AOTAutograd's in-memory cache deduplication for repeated transformer layers. The original `split_module` generates generic node names (`add`, `add_1`) via `_target_to_str`, so structurally identical subgraphs across layers produce identical cache keys. `split_module_simple` was preserving original node names (`add_42`, `add_87`), making each layer appear unique. Fixing `name=None` in `create_node` restored the 78/81 cache hit rate.

**Final result: C++ code was stripped (no measurable benefit), and the 5.5x speedup comes entirely from the lazy GraphModule construction + cache-friendly naming.**

## Test Plan

- `pytest test/test_fx_experimental.py -k split_module_simple -x` — 9 tests pass:
  - `test_split_module_simple_basic` — 2-partition split with structure + correctness check
  - `test_split_module_simple_multi_output` — partition with multiple cross-boundary outputs
  - `test_split_module_simple_single_partition` — degenerate single-partition case
  - `test_split_module_simple_partition_affix` — custom submodule naming
  - `test_split_module_simple_symint_crossing` — SymInt propagation across partitions (torch.compile backend)
  - `test_split_module_simple_equivalence_with_split_module` — output equivalence with original `split_module`
  - `test_split_module_simple_dead_code` — dead code across partitions (migrated from `split_module` tests)
  - `test_split_module_simple_kwargs` — kwargs expansion across partitions (migrated)
  - `test_split_module_simple_noop_graph` — empty graph with no ops (migrated)
- End-to-end: vllm `serve` on 3 models (GLM-4.7-FP8, Llama, GPT) with correct inference output
- Profiled with `TORCH_TRACE` to verify no regression in downstream AOT compilation, inductor compile, or autotuning

### CODE REVIEW

File: test/test_fx_experimental.py [hunk 1]
```diff
@@ -1080,6 +1080,194 @@ def backend(gm, inps):
...
+        mod = Mod()
+        x = torch.randn(3, 4)
+        expected = mod(x)
+
+        traced = torch.fx.symbolic_trace(mod)
+
+        # Split into 2 partitions: relu in 0, sigmoid+tanh in 1
+        node_to_partition = {}
+        counter = 0
+        for node in traced.graph.nodes:
```
Comment (Lucaskabela) [inline test/test_fx_experimental.py]: nit: Refactor this into helper function since we use it so many time

File: torch/fx/passes/split_module.py [hunk 2]
```diff
@@ -681,3 +681,300 @@ def add_placeholder():
...
+def _decompose_size_nodes(m: GraphModule) -> None:
+    """Decompose x.size() into per-dim sym_size.int calls.
+
+    torch.Size objects cannot cross split boundaries because aot_autograd
+    cannot handle them as submodule outputs. This replaces each size() call
+    with individual sym_size.int(x, dim) nodes:
+      - Dynamic dims (SymInt) -> new sym_size.int node
+      - Static dims (plain int) -> inlined as literal constant
+    """
+    size_nodes = list(m.graph.find_nodes(op="call_method", target="size"))
```
Comment (Lucaskabela) [inline torch/fx/passes/split_module.py:712]: Why convert this to a list? If we are just using it to iterate in L697 why not iterate on the find_nodes() output directly?

File: torch/fx/passes/split_module.py [hunk 3]
```diff
@@ -681,3 +681,300 @@ def add_placeholder():
...
+
+    for node in size_nodes:
+        tensor_node = node.args[0]
+        ev = tensor_node.meta.get("example_value")
+        if ev is None:
+            continue
+
+        dims: list[Node | int] = []
+        with m.graph.inserting_after(tensor_node):
+            for i in range(ev.dim()):
```
Comment (Lucaskabela) [inline torch/fx/passes/split_module.py]: could this be an enumerate(ev.shape) to get idx, dim_val directly?

File: torch/fx/passes/split_module.py [hunk 4]
```diff
@@ -681,3 +681,300 @@ def add_placeholder():
...
+                if isinstance(dim_val, torch.SymInt):
+                    dn = m.graph.call_function(
+                        torch.ops.aten.sym_size.int, args=(tensor_node, i)
+                    )
+                    dn.meta["example_value"] = dim_val
+                    dims.append(dn)
+                elif isinstance(dim_val, int):
+                    dims.append(dim_val)
+
+        for user in list(node.users):
```
Comment (Lucaskabela) [inline torch/fx/passes/split_module.py:735]: No need to convert iterable to list here right?

File: torch/fx/passes/split_module.py [hunk 5]
```diff
@@ -681,3 +681,300 @@ def add_placeholder():
...
+@compatibility(is_backward_compatible=False)
+def split_module_simple(
+    m: GraphModule,
+    node_to_partition: dict[Node, int],
+    *,
+    partition_affix: str | None = None,
+) -> GraphModule:
+    """Lightweight graph splitter for simple partition patterns.
+
+    A faster alternative to split_module for graphs without autocast/grad
```
Comment (Lucaskabela) [inline torch/fx/passes/split_module.py]: It would be worth modifying this block comment to explain why these restrictions mean we can use a lighterweight split_module

File: torch/fx/passes/split_module.py [hunk 6]
```diff
@@ -681,3 +681,300 @@ def add_placeholder():
...
+    """
+    # --- Pass 1: Dependency detection + node cloning ---
+
+    # Lazy imports for symbolic shape support — only needed when graph
+    # nodes carry example_value metadata with SymInt/SymFloat values
+    # (i.e. graphs from torch.compile/Dynamo, not symbolic_trace).
+    _sympy = None
+    _free_symbols = None
+
+    def _get_sympy():
```
Comment (Lucaskabela) [inline torch/fx/passes/split_module.py]: nit: this could use typehints too

File: torch/fx/passes/split_module.py [hunk 7]
```diff
@@ -681,3 +681,300 @@ def add_placeholder():
...
+    seen_partitions_set: set[int] = set()
+
+    # --- Pass 1: Dependency detection ---
+
+    for node in m.graph.nodes:
+        # Track SymInt symbol bindings (prefer earlier bindings).
+        val = node.meta.get("example_value")
+        if val is not None and hasattr(val, "node") and hasattr(val.node, "expr"):
+            s0 = val.node.expr
+            if isinstance(s0, _get_sympy().Symbol) and s0 not in symbol_to_node:
```
Comment (Lucaskabela) [inline torch/fx/passes/split_module.py]: I am not sure the lazy import makes sense - seems most paths will trigger these calls so we might be faster to just do the import in this function every time 

Can you measure the difference between just importing vs this lazy one?

File: torch/fx/passes/split_module.py [hunk 8]
```diff
@@ -681,3 +681,300 @@ def add_placeholder():
...
+                if arg is node:
+                    new_args.extend(dims)
+                else:
+                    new_args.append(arg)
+            user.args = tuple(new_args)
+        m.graph.erase_node(node)
+
+
+@compatibility(is_backward_compatible=False)
+def split_module_simple(
```
Comment (Lucaskabela) [inline torch/fx/passes/split_module.py]: IMO this function needs to be a bit simpler - lets split out the methods defined inline to standalone methods and see if they can be shared with split_module to reduce complexity

**Review by Lucaskabela — CHANGES_REQUESTED** [review]

File: torch/fx/passes/split_module.py (hunk 2 above)
Comment (fxdawnn) [reply torch/fx/passes/split_module.py:712]: The reason to do this is because ``aot_autograd`` does not allow tuble to cross boundary so we need special handling for size() information retrieving of tensors. See more information in previous commit in vLLM https://github.com/vllm-project/vllm/pull/36038

Comment (fxdawnn) [reply torch/fx/passes/split_module.py:712]: We mutate the graph during iteration (erase_node, inserting new nodes), so we need to snapshot the list upfront. Can't iterate find_nodes() directly.

File: torch/fx/passes/split_module.py (hunk 8 above)
Comment (fxdawnn) [reply torch/fx/passes/split_module.py]: complexity reduced with helper function. I might also rename this to ``faster_split_module`` or ``simpler_split_module``. Since it's relative. Or maybe the name ``split_module_infer``

File: torch/fx/passes/split_module.py (hunk 7 above)
Comment (fxdawnn) [reply torch/fx/passes/split_module.py]: importing and it does not make a difference in time performance. So I've imported to simplify.

File: torch/fx/passes/split_module.py (hunk 6 above)
Comment (fxdawnn) [reply torch/fx/passes/split_module.py]: updated.

File: torch/fx/passes/split_module.py (hunk 5 above)
Comment (fxdawnn) [reply torch/fx/passes/split_module.py]: Added the tl;dr version of the explanation.

File: torch/fx/passes/split_module.py (hunk 4 above)
Comment (fxdawnn) [reply torch/fx/passes/split_module.py:716]: it was needed as reasons above.

File: torch/fx/passes/split_module.py (hunk 3 above)
Comment (fxdawnn) [reply torch/fx/passes/split_module.py]: yes good idea.

**Comment by Lucaskabela** [comment]
@claude can you explain why we need the convert the find_nodes() to a list?

**Comment by Lucaskabela** [comment]
Okay this confirms my though I had while discussing with @fxdawnn offline - makes sense

File: torch/fx/passes/split_module.py [hunk 9]
```diff
@@ -681,3 +681,286 @@ def add_placeholder():
...
+
+    partition_graphs: dict[int, torch.fx.graph.Graph] = {}
+    partition_env: dict[int, dict[Node, Node]] = defaultdict(dict)
+    partition_inputs: dict[int, dict[str, Node]] = defaultdict(dict)
+    partition_outputs: dict[int, dict[str, Node]] = defaultdict(dict)
+    symbol_to_node: dict[Any, Node] = {}
+    seen_partitions: list[int] = []
+    seen_partitions_set: set[int] = set()
+
+    # --- Pass 1: Dependency detection ---
```
Comment (Lucaskabela) [inline torch/fx/passes/split_module.py]: Let's decompose this function some - putting each pass in isolated functions (defined outside the scope of this one) would help our readability

**Review by Lucaskabela — CHANGES_REQUESTED** [review]
One minor note on the fix but otherwise LGTM

File: torch/fx/passes/split_module.py (hunk 9 above)
Comment (fxdawnn) [reply torch/fx/passes/split_module.py]: restructured! Let me know what you think about it now.

**Review by Lucaskabela — COMMENTED** [review]
With the decomposition this looks good to me :) Want to let @zou3519 take a look as well before approving

**Comment by zou3519** [comment]
@Lucaskabela this PR is a bit too big for me to review right now. If you think it's good feel free to ship it, but someone needs to do the enablement work of making sure this works in vLLM and owning the success/failure of it. It sounds like Xiao doesn't have bandwidth to work more on this after shipping this PR, so if you want to take it over from here feel free to, but you don't have to.

**Comment by Lucaskabela** [comment]
@claude

**Review by Lucaskabela — CHANGES_REQUESTED** [review]
Claude nits seem reasonable - let us fix these

**Comment by Lucaskabela** [comment]
@claude

**Comment by Lucaskabela** [comment]
@fxdawnn can we apply these?

**Review by Lucaskabela — CHANGES_REQUESTED** [review]
Still need to apply some suggestions missed from last round

**Comment by fxdawnn** [comment]
All three are in hot paths during compilation. The size(dim) check adds a branch on every size node, the .get() adds overhead on every symbol lookup, and none of them fix an actual observed bug.

But I'd apply those. Are we still planning on shipping to vLLM or not?

**Review by Lucaskabela — APPROVED** [review]
Thanks for the change - lgtm! we can worry about vllm integration later
